### PR TITLE
Add support for extension.json

### LINKED ISSUE #126: Add support for MWs new extension registration
This needs to be done in such a way the extension still works with older versions of MW (down to MW 1.23) and continues to work via Composer in the way existing users expect it to.

Some of the features of the extension registration should also not be used:
- Hooks
- autoloading
- SpecialPages, Actions, JobClasses, APIModules

This one is up for the taking. Help with it is much appreciated.

### PR DESCRIPTION
**Pull request by paladox**
Fixes #126

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
Needs rebase

**Comment by paladox** [comment]
Ok rebased.

Probaly needs some testing by you since you would know what to expect please.

**Comment by paladox** [comment]
Im not sure why it fails with https://travis-ci.org/JeroenDeDauw/Maps/jobs/135067644

**Comment by JeroenDeDauw** [comment]
Thanks for working on this.

It fails because this code did not get executed

```
 	$GLOBALS['wgParamDefinitions']['coordinate'] = [
 		'string-parser' => 'DataValues\Geo\Parsers\GeoCoordinateParser',
 		'string-parser' => 'DataValues\Geo\Parsers\GeoCoordinateParser',
 	];
```

I probably should have specified this more clearly in the ticket: I do not want references to PHP constructs, be it class names, or function names, in the new extension registration thing. IMO it's a bad idea to support those there, and I'm sad it got implemented like that in MediaWiki.

**Comment by paladox** [comment]
Oh how should add it. Should we do it through a callback or the extension function.

Do you have an example on how you would like it done please.

**Comment by JeroenDeDauw** [comment]
I don't care so much where this is put, as long as it is in PHP space, and it works. It's easy enough to move it around anyway. Btw, I recommend using `::class` to reference class names without using strings. So instead of

```
	$GLOBALS['wgParamDefinitions']['coordinate'] = [
		'string-parser' => '\DataValues\Geo\Parsers\GeoCoordinateParser',
	];
```

You'd have

```
	$GLOBALS['wgParamDefinitions']['coordinate'] = [
		'string-parser' => \DataValues\Geo\Parsers\GeoCoordinateParser::class,
	];
```

Like this things don't break if someone renamed the class with a proper tool.

**Comment by JeroenDeDauw** [comment]
Needs rebase again

@paladox is this something you want to continue with?

**Comment by kghbln** [comment]
Ping @paladox

**Comment by kghbln** [comment]
@paladox No worries if you do not like to continue on this. Please just say so if this is the case ...

**Comment by paladox** [comment]
Hi, please completely forgot about this. I had a tab open for this but something else came up and then I forgot about this.

**Comment by kghbln** [comment]
No worries. I know such situations very well. So you are continuing on this?

**Comment by paladox** [comment]
I rebased now.

File: Maps.php [hunk 1]
```diff
@@ -109,40 +127,19 @@
...
+
+	$GLOBALS['wgHooks']['ParserFirstCallInit'][] = 'MapsHooks::onParserFirstCallInit3';
+
+	$GLOBALS['wgHooks']['ParserFirstCallInit'][] = 'MapsHooks::onParserFirstCallInit4';
+
+	$GLOBALS['wgHooks']['ParserFirstCallInit'][] = 'MapsHooks::onParserFirstCallInit5';
+
+	$GLOBALS['wgHooks']['ParserFirstCallInit'][] = 'MapsHooks::onParserFirstCallInit6';
+
+	$GLOBALS['wgHooks']['ParserFirstCallInit'][] = 'MapsHooks::onParserFirstCallInit7';
```
Comment (JeroenDeDauw) [inline Maps.php]: Not sure why you did this replace. I like the older form much more.

File: extension.json [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+		],
+		"ParserFirstCallInit": [
+			"MapsHooks::onParserFirstCallInit1",
+			"MapsHooks::onParserFirstCallInit2",
+			"MapsHooks::onParserFirstCallInit3",
+			"MapsHooks::onParserFirstCallInit4",
+			"MapsHooks::onParserFirstCallInit5",
+			"MapsHooks::onParserFirstCallInit6",
+			"MapsHooks::onParserFirstCallInit7"
+		],
```
Comment (JeroenDeDauw) [inline extension.json]: This is now both here and in `Maps.php`

File: extension.json [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+			},
+			"distance": {
+				"string-parser": "Maps\\DistanceParser::class"
+			},
+			"wmsoverlay": {
+				"string-parser": "Maps\\WmsOverlayParser::class"
+			},
+			"mapsimageoverlay": {
+				"string-parser": "Maps\\ImageOverlayParser::class"
+			},
```
Comment (JeroenDeDauw) [inline extension.json]: As I wrote before in the thread, please leave this out of the JSON file

File: Maps.hooks.php [hunk 4]
```diff
@@ -10,6 +10,123 @@
...
+			throw new Exception( 'You need to have Validator installed in order to use Maps' );
+		}
+
+		$GLOBALS['egMapsStyleVersion'] = $GLOBALS['wgStyleVersion'] . '-' . Maps_VERSION;
+	}
+
+	public static function onExtensionFunction() {
+		if ( $GLOBALS['egMapsDisableExtension'] ) {
+			return true;
+		}
```
Comment (JeroenDeDauw) [inline Maps.hooks.php]: This is now duplicated with code in `Maps.php`

**Comment by JeroenDeDauw** [comment]
Something was broken with these changes as shown by the tests: https://travis-ci.org/JeroenDeDauw/Maps/jobs/238720483

File: Maps.php (hunk 1 above)
Comment (paladox) [inline Maps.php]: Needs to be done otherwise we would have duplicate code and would need to update it in two places.

File: extension.json (hunk 2 above)
Comment (paladox) [inline extension.json]: Yes, to keep combat with mw 1.23.

File: Maps.hooks.php (hunk 4 above)
Comment (paladox) [inline Maps.hooks.php]: Yep has to be done since users can bypass Maps.php and load extension.json directly.

File: extension.json (hunk 3 above)
Comment (paladox) [inline extension.json]: Thanks, done

**Comment by paladox** [comment]
Im not sure how this change breaks the tests.

File: extension.json (hunk 2 above)
Comment (JeroenDeDauw) [inline extension.json]: Why not have this only in `Maps.php`? I don't see any advantage of having it in the JSON file

File: Maps.hooks.php (hunk 4 above)
Comment (JeroenDeDauw) [inline Maps.hooks.php]: Sounds like this again is for compat reasons. You can drop support for MW < 1.27.

Comment (JeroenDeDauw) [inline Maps.hooks.php]: Hah, I just realized that MW support for 1.23 ended 2 days ago anyway

File: extension.json (hunk 2 above)
Comment (paladox) [inline extension.json]: As we can remove support for 1.27 and lower we can now keep this in extension.json. Also users can overide Maps.php by loading extension.json only. php entry points are also deprecated.

File: Maps.hooks.php (hunk 4 above)
Comment (paladox) [inline Maps.hooks.php]: Yep

**Pull request closed without merging**